### PR TITLE
adding stat module in fdg

### PR DESCRIPTION
**Pull request by MoodyMudit**

### CODE REVIEW

File: hubblestack/extmods/fdg/stat.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    if 'allow_more_strict' in expected.keys() and 'mode' not in expected.keys():
+        reason = "'allow_more_strict' tag can't be specified without 'mode' tag." \
+                 " Seems like a bug in hubble profile."
+        ret = {'Failure': reason, "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
+        return False, ret
+
+    return True, ret
+
+def _check_mode(max_permission, given_permission, allow_more_strict):
```
Comment (goravsingal) [inline hubblestack/extmods/fdg/stat.py:181]: This is exactly duplicate of stat module code.

We should put this code at a single place. In some common file. (utils/fs_commons.py)
And, should use that at both places.

File: hubblestack/extmods/fdg/stat.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        if attribute == 'allow_more_strict' or attribute == 'match_on_file_missing':
+            continue
+        file_attribute_value = salt_ret[attribute]
+
+        if attribute == 'mode':
+            if file_attribute_value != '0':
+                file_attribute_value = file_attribute_value[1:]
+            allow_more_strict = False
+            if 'allow_more_strict' in expected.keys():
+                allow_more_strict = expected['allow_more_strict']
```
Comment (devagupt) [inline hubblestack/extmods/fdg/stat.py:111]: can be combined in a single line
`allow_more_strict = expected.get('allow_more_strict', False)`

File: hubblestack/extmods/fdg/stat.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+import salt.utils
+import salt.utils.platform
+import os
+log = logging.getLogger(__name__)
+
+
+def check_stats(params='', chained=None, chained_status=None):
+    '''
+
+    :param params:
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done

File: hubblestack/extmods/fdg/stat.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    :param chained:
+    :param chained_status:
+    :return: tuple with (status(Boolean), result(dict))
+    This function takes into input stat params that are to be matched with stats of given file.
+    The filepath can be provided either directly in params or through chaining. See example above.
+    '''
+    params = params.get('params')
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Handled this.

File: hubblestack/extmods/fdg/stat.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    module: osquery.query       # this tells FDG to call osquery submodule's query function
+    args:
+       - 'select path as filepath from FILE where path="/etc/docker/daemon.json";'
+    xpipe:                      # This tells FDG to pass the output of the above query to check_stats functions of this profile
+       check_stats
+
+check_stats:                    # this profile's check_stats module
+    module: stat.check_stats   # this tells FDG to call stat submodule's check_stats function
+    args:
+        - params:
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Removed this.

File: hubblestack/extmods/fdg/stat.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                 "File might not exist, or hubble might not" \
+                 " have enough permissions".format(filepath)
+        ret = {'Failure': reason, "expected": expected}
+        log.info("FDG stat is returning status : False, value : {0}".format(ret))
+        return False, ret
+
+
+def _validate_inputs(filepath, expected):
+    ret = ''
+    if filepath is None or filepath == '':
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:180]: does it cover both cases in that way?

File: hubblestack/extmods/fdg/stat.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        log.error("filepath not specified")
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:181]: The logger will tell us from which file this log is coming. Can you please add what param can I add to make this more specific.

File: hubblestack/extmods/fdg/stat.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: done.

File: hubblestack/extmods/fdg/stat.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    :param params:
+    :param chained:
+    :param chained_status:
+    :return: tuple with (status(Boolean), result(dict))
+    This function takes into input stat params that are to be matched with stats of given file.
+    The filepath can be provided either directly in params or through chaining. See example above.
+    '''
+    params = params.get('params')
+
+    if chained != None:
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:73]: Done.

File: hubblestack/extmods/fdg/stat.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+                reason = {'expected': str(expected[attribute]),
+                          'current': str(file_attribute_value)}
+                reason_dict[attribute] = reason
+
+    if reason_dict:
+        ret = {'Failure': "For file '{0}': {1}".format(filepath, reason_dict), "expected": expected}
+
+    if passed:
+        ret = {'Success': 'all stats matching', "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:140]: Added filepath in Success log.

File: hubblestack/extmods/fdg/stat.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:180]: done.

File: hubblestack/extmods/fdg/stat.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+def _validate_inputs(filepath, expected):
+    ret = ''
+    if filepath is None or filepath == '':
+        log.error("filepath not specified")
+        ret = {'Failure': "no filepath provided", "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
+        return False, ret
+
+    if 'allow_more_strict' in expected.keys() and 'mode' not in expected.keys():
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:186]: Done.

File: hubblestack/extmods/fdg/stat.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        log.info("FDG stat is returning status : False, value : {0}".format(ret))
+        return False, ret
+
+
+def _validate_inputs(filepath, expected):
+    ret = ''
+    if filepath is None or filepath == '':
+        log.error("filepath not specified")
+        ret = {'Failure': "no filepath provided", "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done.

File: hubblestack/extmods/fdg/stat.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+        return False, ret
+
+    if 'allow_more_strict' in expected.keys() and 'mode' not in expected.keys():
+        reason = "'allow_more_strict' tag can't be specified without 'mode' tag." \
+                 " Seems like a bug in hubble profile."
+        ret = {'Failure': reason, "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done.

File: hubblestack/extmods/fdg/stat.py (hunk 2 above)
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:112]: Done.

File: hubblestack/extmods/fdg/stat.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+
+    log.debug("file stats are {0}".format(salt_ret))
+    if not salt_ret:
+        log.info("file stats couldn't be fetched, checking corner cases")
+        return _check_corner_cases(filepath, expected)
+
+    passed = True
+    reason_dict = {}
+    for attribute in expected.keys():
+        if attribute == 'allow_more_strict' or attribute == 'match_on_file_missing':
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:103]: Is this better over the current implementation?

File: hubblestack/extmods/fdg/stat.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+    log.info("checking stats of {0}".format(filepath))
+    if os.path.exists(filepath):
+        salt_ret = __salt__['file.stats'](filepath)
+    else:
+        salt_ret = {}
+
+    log.debug("file stats are {0}".format(salt_ret))
+    if not salt_ret:
+        log.info("file stats couldn't be fetched, checking corner cases")
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py:97]: Added filepath

File: tests/unittests/test_fdg_stat.py [hunk 15]
```diff
@@ -0,0 +1,400 @@
+# coding: utf-8
+import logging
+import hubblestack.extmods.fdg.stat
+import mock
+log = logging.getLogger(__name__)
+
+
+def test_validate_inputs_positive():
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply tests/unittests/test_fdg_stat.py:8]: The names of test cases are descriptive themselves.

File: hubblestack/extmods/fdg/stat.py [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+    This function takes into input stat params that are to be matched with stats of given file.
+    The filepath can be provided either directly in params or through chaining. See example above.
+    '''
+
+    if params == '' or params is None:
+        ret = {'Failure' : 'invalid input, no params provided'}
+        return False, ret
+
+    if chained:
+        log.info("value of 'chained' is not null, using {0} value as filepath".format(chained))
```
Comment (jettero) [inline hubblestack/extmods/fdg/stat.py:74]: Do not use format before passing to log.info ... log.info already uses the older style formatting rules so this is somewhat redundant and the interpolation may not be necessary.

`log.info("value of 'chained' is not null, using %s value as filepath", chained)`

File: hubblestack/extmods/fdg/stat.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        log.info("Invalid inputs provided in fdg stat module, returning False")
+        return False, ret
+
+    log.info("checking stats of {0}".format(filepath))
+    if os.path.exists(filepath):
+        salt_ret = __salt__['file.stats'](filepath)
+    else:
+        salt_ret = {}
+
+    log.debug("file stats are {0}".format(salt_ret))
```
Comment (jettero) [inline hubblestack/extmods/fdg/stat.py:95]: No format, please

File: hubblestack/extmods/fdg/stat.py (hunk 1 above)
Comment (jettero) [reply hubblestack/extmods/fdg/stat.py:195]: @goravsingal agree.

File: tests/unittests/test_fdg_stat.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jettero) [reply tests/unittests/test_fdg_stat.py:8]: I think the expected outcome is best documented as assert statements, personally. But certainly an overview of what we're testing is desirable.

[the remark this answers is not in the document]
Comment (jettero) [reply tests/unittests/test_fdg_stat.py:8]: I think I also agree that the name of the test case can describe the test just fine. Extra comments for the sake of adding comments doesn't really help anything. It's just more stuff you have to update if you make changes.

File: tests/unittests/test_fdg_stat.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+    hubblestack.extmods.fdg.stat._validate_inputs = mock.Mock(return_value=valid_inputs)
+    val = hubblestack.extmods.fdg.stat.check_stats(params=params)
+    assert not val[0]
+    assert isinstance(val[1], dict)
+    assert 'Failure' in val[1].keys()
+    assert 'expected' in val[1].keys()
+
+
+def test_check_stats_negative_no_params():
+    log.info('\n \n Executing test_check_stats_negative_no_params')
```
Comment (jettero) [inline tests/unittests/test_fdg_stat.py:227]: I understand the desire to add some vertical separation to the output, but please don't put newlines in log lines.

File: tests/unittests/test_fdg_stat.py [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+
+
+def test_check_mode_positive_allow_more_strict():
+    log.info('\n \n Executing test_check_mode_positive_allow_more_strict')
+    val = hubblestack.extmods.fdg.stat._check_mode("644", "600", True)
+    assert val
+
+
+def test_check_mode_negative_allow_more_strict():
+    log.info('\n \n Executing test_check_mode_negative_allow_more_strict')
```
Comment (jettero) [inline tests/unittests/test_fdg_stat.py:405]: please don't put newlines in log lines. I know they're just test logs, but sometimes I view test logs with `--log-cli-level whatever` and it'd just be that much more extra noise.

File: hubblestack/extmods/fdg/stat.py [hunk 20]
```diff
@@ -0,0 +1,264 @@
+'''
```
Comment (jettero) [inline hubblestack/extmods/fdg/stat.py:1]: I think even the module doc should start with `"""` rather than `'''` … it's a pep8 rule iirc. I don't completely agree with it, personally, but it's worth conforming to something for consistency's sake.

File: tests/unittests/test_fdg_stat.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (devagupt) [reply tests/unittests/test_fdg_stat.py:8]: @MoodyMudit some times the names of test cases becomes too long to make them meaningful.
For instance we have a test case named `test_validate_inputs_negative_no_mode` here in this file which is not so small. 
I prefer to use a generic name as `test_validate_inputs_1` with proper docstring mentioning the purpose of this test case and expected behavior but I agree that naming convention is best decided by the author of that code.
Apart from function naming, pylint still requires us to add a docstring to each function whether be it unit test or not. So, docstrings are still needed.

[the remark this answers is not in the document]
Comment (MoodyMudit) [reply tests/unittests/test_fdg_stat.py:8]: I don't think having generic function/test case names is a good idea. This makes it difficult to refer to a particular thing unless you have that opened in front on you. A few months from now if anyone refers `test_validate_inputs_1` to me, I won't have any context of what is being asked. However, the test case name (if given descriptive) would give me that context.

File: hubblestack/extmods/fdg/stat.py (hunk 1 above)
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Moved to a new util. Changed the code at both the places. Also moved the corresponding test cases to new file.

File: tests/unittests/test_fdg_stat.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply tests/unittests/test_fdg_stat.py:8]: Added doc strings.

File: hubblestack/extmods/fdg/stat.py (hunk 16 above)
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done.

File: hubblestack/extmods/fdg/stat.py (hunk 17 above)
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done.

File: tests/unittests/test_fdg_stat.py (hunk 18 above)
Comment (MoodyMudit) [reply tests/unittests/test_fdg_stat.py]: Removed them.

File: tests/unittests/test_fdg_stat.py (hunk 19 above)
Comment (MoodyMudit) [reply tests/unittests/test_fdg_stat.py]: Removed them.

File: hubblestack/extmods/fdg/stat.py (hunk 20 above)
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Done.

File: hubblestack/extmods/fdg/stat.py [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+        return True, ret
+    else:
+        log.info("FDG stat is returning status : False, value : {0}".format(ret))
+        return False, ret
+
+
+def _check_corner_cases(filepath, expected):
+    if not expected:
+        ret = {"Success" : "nothing is expected, therefore passing the test", "expected": expected}
+        log.info("FDG stat is returning status : True, value : {0}".format(ret))
```
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Added more specific items like function name to logs.

File: hubblestack/extmods/fdg/stat.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: I ran pylint. Score is 9.89. The only complain is 
`W: 59,41: Unused argument 'chained_status' (unused-argument)`

**Comment by mew1033** [comment]
While I think the code here looks good, I think we're using fdg here in a sub-optimal way. Fdg is meant do gather data, not report a success or failure. IMO, adding a stat module to fdg would mean that fdg can now run stat on a file and report all of the resulting attributes to Splunk. If we want to use the fdg<--> nova connector to report success or failure, maybe we should have another fdg function that can take the output of something (in this case stat), and compare it to known good/bad values and report success/failure.
Thoughts?

**Comment by MoodyMudit** [comment]
@mew1033 Yes I see the problem that you addressed. An FDG module should be used for data gathering as well.
So, I thought of a comparator function, but that comparator would only be used with file stats because file permissions are compared very uniquely and so this comparator would be of no use with data other than file stats. Also we already have grep to 'compare' the output of something and checking if expected output was found or not. Having said this, I would want to have the comparator function inside the stat module only, because, again, file stats(permissions) comparison is a very unique kind of comparison.
Now, we have to make this stat module to work both ways, as a data collector and also as a final comparator (when invoked from nova, where expected output is a TRUE/FALSE).
How about we have an additional boolean flag, that tells what the output is supposed to be? It can then either collect data and send to splunk (when invoked as an FDG module) or it could fetch actual file stats and compare them to given values, finally giving a PASS/FAIL status (when invoked as a NOVA module).

File: hubblestack/extmods/fdg/stat.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (goravsingal) [reply hubblestack/extmods/fdg/stat.py]: Yes we will come to know which line. But, if this line came from more than 1 times, we will not be sure for which file or check this is.
We can add file, or check_id, or tag whatever we have at this place.

[the remark this answers is not in the document]
Comment (MoodyMudit) [reply hubblestack/extmods/fdg/stat.py]: Ohh...now I get your point. Good thought. Thanks.

**Comment by mew1033** [comment]
@MoodyMudit That makes sense. So it sounds like the util functions should exist to compare values and return a boolean. Then we should have different functions available in the fdg stat module:

1. Run stat and get the details about a file. Pass all these details into the following module.
2. Run stat and check a specific value (permissions for example) against a known good value. This known good value can come from the fdg routine itself, or possibly be passed in from the previous module. This function returns either true or false and is then a great candidate for feeding into the nova<-->fdg connector.

Only the second function calls out to the stat_functions util file.
